keps/sig-node/2712-pod-priority-based-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,553 @@
+of name. 
+
+Instead of looking at the pod priority class names, we can instead
+look at the pod priority class values to allow more control over pod
+shutdown order.
+
+### Goals
+
+*   Make the kubelet use shutdown configuration based on pod priority values for
+    graceful shutdown.
+
+### Non-Goals
+
+*   Non-Linux hosts aren't supported
+*   Let users modify or change existing pod lifecycle or introduce new inner
+    pod depencides / shutdown ordering
+*   Provide guarantee to handle all cases of graceful node shutdown, for
+    example abrupt shutdown or sudden power cable pull can’t result in graceful
+    shutdown
+
+## Proposal
+
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+*   As a cluster administrator, I can configure the nodes in my cluster to
+    allocate different graceful shutdown durations for different pod priority
+    value ranges to terminate them gracefully during node shutdown
+
+
+### Implementation
+
+This implementation builds on top of the node graceful shutdown feature
+by introducing additional configuration. A new feature flag called
+`PodPriortityBasedGracefulShutdown` will be added to control the behavior
+of the kubelet.
+
+We will describe the configuration by using an example. Say, the
+following custom pod priority classes are created in a cluster:
+
+|Pod priority class name|Pod priority class value|
+|-----------------------|------------------------|
+|custom-class-a         | 100000                 |
+|custom-class-b         | 10000                  |
+|custom-class-c         | 1000                   |
+|regular/unset          | 0                      |
+
+We could set kubelet configuration to stop the pods as:
+
+|Pod priority class value|Shutdown period|
+|------------------------|---------------|
+| 100000                 |300 seconds    |
+| 10000                  |180 seconds    |
+| 1000                   |120 seconds    |
+| 0                      |60 seconds     |
+
+The above table implies that any pod with priority value >= 100000 will get
+300 seconds to stop, any pod with value >= 10000 and < 100000 will get 180
+seconds to stop, any pod with value >= 1000 and < 10000 will get 120 seconds to stop. 
+Finally, all other pods will get 60 seconds to stop.
+
+Note: We use priority values instead of names because k8s API design discourages
+using names and the values are more portable as well.
+
+One doesn't have to specify values corresponding to all of the classes. For
+e.g. the config could also be
+
+|Pod priority class value|Shutdown period|
+|------------------------|---------------|
+| 100000                 |300 seconds    |
+| 1000                   |120 seconds    |
+| 0                      |60 seconds     |
+
+
+In the above case, the pods with custom-class-b will go into the same bucket
+as custom-class-c for shutdown.
+
+If there are no pods in a particular range, then the kubelet will skip
+over the range entirely.
```

Comment:
What does “skip over the range entirely” mean? 0-second grace period perhaps?

Reply:
If we allocate 2 minutes for pods in a 1000 - 10000 range and there are no actual pods then we don't artificially wait for 2 minutes but move on to the next range.